docs/simplified_security_model.md

```diff
@@ -61,42 +61,48 @@ authentication is described below.
 By authorized trust it is meant that the Target is granted the necessary
 credentials (by the Client) to be able to identify Peers as genuine (using
-whatever service uses these Certificates). The Target is authorized trust by the
+whatever service uses these certificates). The Target is authorized trust by the
 Client via the successful execution of either:
 
 *   gNOI certificate installation or rotation under a Secure connection (where
     the Target is not compromised), or
 *   other provisioning or bootstrapping mechanism described outside of the scope
     of this security model.
 
-## Target Certificate installation or rotation
+## Target certificate installation or rotation
 
 To initiate certificate installation using the gNOI Certificate Management
 service:
 
 1.  The Client sends parameters to the target so that it can generate a
-    Certificate Signing Request (CSR).
+    certificate Signing Request (CSR).
 2.  The Target then generates this CSR using its Private Key.
-3.  The Target then replies with the Certificate to the Client, which
-    subsequently provides the Certificate to the CA to be signed.
-4.  The signed Certificate is then returned to the Target by the Client and it
-    becomes one of the Target’s Certificates.
-
-Each one of the Target’s Certificate has a unique identifier, the Certificate
-ID. This ID is used both to install and to rotate a Certificate. Certificate
-installation for an existing Certificate ID MUST fail. Replacing an existing
-Certificate must rather use the Certificate rotation mechanism, whose steps are
+3.  The Target then replies with the certificate to the Client, which
+    subsequently provides the certificate to the CA to be signed.
+4.  The signed certificate is then returned to the Target by the Client and it
+    becomes one of the Target’s certificates.
+
+Each one of the Target’s certificate has a unique identifier, the certificate
+ID. This ID is used both to install and to rotate a certificate. certificate
+installation for an existing certificate ID MUST fail. Replacing an existing
+certificate must rather use the certificate rotation mechanism, whose steps are
 similar to the ones described above with the addition of final validation.
 
 ## Target CA pool
 
-In order for the Target to validate the Client or a Peer's Certificates, it must
-have a pool of one or more CA Certificates. These are provisioned onto the
-Target by the Client during creation or rotation of the Target’s Certificates.
+In order for the Target to validate the Client or a Peer's certificates, it must
+have a pool of one or more CA certificates. These are provisioned onto the
+Target by the Client during creation or rotation of the Target’s certificates.
+
+### Validate installed certificate (optional)
+
+For increased security, the Target may use certificates in its CA pool to
+validate a newly installed certificate. This requires the CA pool to contain a
+CA certificate that can validate the new certificate.
```

Comment:
I agree that the performance of 1) has been helped by https://github.com/openconfig/gnoi/pull/22 (assuming the suggested order is followed). 

For 2 b) when a client presents a certificate for a new session, we know a priori whether the client certificate has to be validated and if so by which CA-Cert (it has been installed by the bundle in a previous install).
For 2 a) when a new certificate is installed, since we don't know which certificate in the bundle ca_certificates should be used to validate the certificate being installed, then the worse case is that we need to try all the certificates in the bundle.

Reply:
for 2b) If the CA bundle has a few Certificates, how do you know which one of them to use for validating the client? If there is a direct bind, I don't know about it.

Reply:
Sam, you can figure this out at install time. So if your bundle has Int1-CA-Cert, Int2-CA-Cert, Root-CA-Cert, by validating the signatures you can know which one is the "lowest" CA-cert which should be used to validate the client cert when it is presented.
 
Did I misunderstand your question?

Reply:
I am looking to understand what is the difference in the method for validating any session certificate versus the method for validating a newly installed certificate. I am assuming they are the exact same method. Both attempt to validate against all the "lowest" CA-certs in the bundle.

Reply:
Let's start in default state.

(i) When you install the first certificate, the bundle has the CA-certs to authenticate the client subsequently. If the bundle has the CA-certs in an expected order (as in a certificate-chain), the target knows the 1st one is the "lowest" CA cert, validates it using the 2nd entry etc until the last CA-cert which is the Root CA-cert. If the bundle has the CA-certs in an unknown/unexpected order, then the target has to figure out which one is the lowest and the next one etc until the Root CA-Cert. With 2-3 CA-certs in the bundle, not a big deal. With says 10 certs, it becomes a big deal to figure out the lowest and validate all the CA-Certs in the chain (lots of combinations). This is why we had discussed a "expected order" for the certs and you added that in a recent commit. 

(ii) Let's say you install a second certificate (we're now in provisioned state), we have mutual-auth so the client presents a certificate to the target. We already know from (i) which is the lowest CA-cert which should be used to authenticate the client, so we don't need to look for it again, we don't have the same hit as in (i). Same thing for subsequent RPCs until a new bundle is provided.

(iii) To validate the target certificate being installed, the target also has to try all CA-certs in the bundle (unless it knows which one should be used).

Reply:
Got it. So there is always an extra effort whenever the CA bundle is provided. For situations where the CA bundle is empty the system already has a mapping for the chains in the stored CA list.

Reply:
Yes there is extra work when the CA bundle is provided. This is why we'd like some "predictability" for the bundle: expected order of the CA-certs for peer authentication and location of the CA-cert to validate the target cert being installed. Thanks.

Reply:
Sam, your comment above about CA bundle empty reminded me of a question which I think I never asked: when the CA bundle is empty does it mean continue using the last CA bundle which was provided or does it mean delete the CA bundle (no more peer authentication). I believe it's the former (please confirm), if so then does that mean that the CA bundle is never deleted?

Reply:
Reshad, the documentation states that when not empty, the provided certificates should squash the existing bundle, see:
https://github.com/openconfig/gnoi/blob/master/cert/cert.proto#L277

The only way to delete the CA bundle is to factory reset the device:
https://github.com/openconfig/gnoi/pull/17/files

Reply:
Sam, thanks for confirming that empty CA bundle means no change.